{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMaSgmFlJBCQ883oAGuv6ak"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Excepciones"],"metadata":{"id":"Wbj9M1v4aAoF"}},{"cell_type":"code","source":["## Para controlar cualquier error que pueda ocurrir durante la ejecución de un programa"],"metadata":{"id":"WTkoFrjrgY7M"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def intdiv(a: int, b: int) -> int:\n","    return a // b"],"metadata":{"id":"oUXwVALzZ-pk","executionInfo":{"status":"ok","timestamp":1706484989586,"user_tz":300,"elapsed":6,"user":{"displayName":"Jhon Cristian Jimenez Mariño","userId":"13494194301674030293"}}},"execution_count":1,"outputs":[]},{"cell_type":"code","source":["intdiv(3, 0)"],"metadata":{"id":"8NiMmTMfZ-sT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["## Para manejar las excepciones podemos usar un bloque de código con las palabras reservadas try and except:\n","## Si se produce una excepción, ésta será capturada por el bloque except, ejecutándose el código que contiene.\n","def intdiv(a: int, b: int) -> int:\n","    try:\n","        return a // b\n","    except:\n","        print('Please do not divide by zero...')"],"metadata":{"id":"43IkFJR0Z-u_","executionInfo":{"status":"ok","timestamp":1706485087664,"user_tz":300,"elapsed":5,"user":{"displayName":"Jhon Cristian Jimenez Mariño","userId":"13494194301674030293"}}},"execution_count":5,"outputs":[]},{"cell_type":"markdown","source":["# Especificando Excepciones"],"metadata":{"id":"cClgC6L2dEhs"}},{"cell_type":"code","source":["## En el siguiente ejemplo mejoraremos el código anterior, capturando distintos tipos de excepciones predefinidas:\n","\n","## TypeError: Operación sobre un objeto de tipo inapropiado                        Ejemplo:  'x' / 3\n","## IndexError: Subíndice de secuencia fuera de rango                               Ejemplo:  (2, 3)[5]\n","## ValueError: Operación sobre un objeto de tipo correcto pero valor inapropiado   Ejemplo:  int('x')\n","## ZeroDivisionError: Segundo argumento de división o módulo es cero               Ejemplo:  1 / 0\n","## KeyError: Clave de diccionario no encontrada                                    Ejemplo:  {0:1, 1:2}[2]\n","## AttributeError: Referencia de atributo inexistente                              Ejemplo:  'hello'.splik()"],"metadata":{"id":"PG3UN7ggflF3"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def intdiv(a, b):\n","    try:\n","        result = a // b\n","    except TypeError:\n","        print('Check operands. Some of them seems strange...')\n","    except ZeroDivisionError:\n","        print('Please do not divide by zero...')\n","    except Exception:\n","        print('Ups. Something went wrong...')"],"metadata":{"id":"1wo-y84rZ-x2","executionInfo":{"status":"ok","timestamp":1706485548920,"user_tz":300,"elapsed":256,"user":{"displayName":"Jhon Cristian Jimenez Mariño","userId":"13494194301674030293"}}},"execution_count":6,"outputs":[]},{"cell_type":"code","source":["intdiv(3, 0)\n","intdiv(3, '0')"],"metadata":{"id":"KI-KkcOtdDoR"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Agrupando Excepciones"],"metadata":{"id":"1NJpKNZzgQmM"}},{"cell_type":"code","source":["## Si nos interesa tratar distintas excepciones con el mismo comportamiento, es posible agruparlas en una única línea:"],"metadata":{"id":"K13AmYx-dDrQ","executionInfo":{"status":"ok","timestamp":1706486342523,"user_tz":300,"elapsed":6,"user":{"displayName":"Jhon Cristian Jimenez Mariño","userId":"13494194301674030293"}}},"execution_count":8,"outputs":[]},{"cell_type":"code","source":["def intdiv(a, b):\n","    try:\n","        result = a // b\n","    except (TypeError, ZeroDivisionError):\n","        print('Check operands: Some of them caused errors...')\n","    except Exception:\n","        print('Ups. Something went wrong...')"],"metadata":{"id":"oraZhstSdDuo","executionInfo":{"status":"ok","timestamp":1706486343824,"user_tz":300,"elapsed":227,"user":{"displayName":"Jhon Cristian Jimenez Mariño","userId":"13494194301674030293"}}},"execution_count":9,"outputs":[]},{"cell_type":"code","source":["intdiv(3, 0)"],"metadata":{"id":"mEJuvyhgZ-z7"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Variantes en el Tratamiento de Excepciones"],"metadata":{"id":"Gm_NnCbEhXc9"}},{"cell_type":"code","source":["## La Clausula else es para saber que todo ha ido bien y que no se ha lanzado ninguna excepción. (Importante)\n","## La cláusula finally que se ejecuta siempre, independientemente de si ha habido o no ha habido error."],"metadata":{"id":"RAgSh1eQZ-3X"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["values = [4, 2, 7]\n","\n","try:\n","    r = values[3]\n","except IndexError:\n","    print('Error: Index not in list')\n","else:\n","    print(f'Your wishes are my command: {r}')\n","finally:\n","    print('Have a good day!')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"nAqk0qnZhWue","executionInfo":{"status":"ok","timestamp":1706486637373,"user_tz":300,"elapsed":4,"user":{"displayName":"Jhon Cristian Jimenez Mariño","userId":"13494194301674030293"}},"outputId":"a4814b73-8850-4b4d-d064-dffaf2600b9f"},"execution_count":13,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: Index not in list\n","Have a good day!\n"]}]},{"cell_type":"markdown","source":["# Mostrando los Errores"],"metadata":{"id":"xwq9yJUgjaOw"}},{"cell_type":"code","source":["## Podemos mostrar sus mensajes de error asociados. Para ello tendremos que hacer uso de la palabra reservada as\n","## junto a un nombre de variable que contendrá el objeto de la excepción."],"metadata":{"id":"YIPUgXGAhWxH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["try:\n","    print(values[3])\n","except IndexError as err:\n","    print(err)"],"metadata":{"id":"ZXMWFLzjhWza"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["## Una vez con la excepción capturada, ya podemos «elaborar» un poco más el mensaje de salida:\n","try:\n","    print(values[3])\n","except IndexError as err:\n","    print(f'Something went wrong: {err}')"],"metadata":{"id":"DN_YeUp_jgFk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"JyVKwlvnjgIC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"10b4TzBdjgLf"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"E4F8cJcZhW10"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"DyUdpEMmhW5P"},"execution_count":null,"outputs":[]}]}